src/vcpkg/sourceparagraph.cpp

```diff
@@ -708,143 +788,252 @@ namespace vcpkg
+                        next();
+                        break;
+                    default:
+                        if (cur() > 0x7F)
+                        {
+                            auto ch = cur();
+                            auto first = it().pointer_to_current();
+                            next();
+                            auto last = it().pointer_to_current();
+                            add_error(msg::format(msgLicenseExpressionContainsUnicode,
+                                                  msg::value = ch,
+                                                  msg::pretty_value = StringView{first, last}));
+                            break;
+                        }
+                        if (!is_idstring_element(cur()))
+                        {
+                            add_error(msg::format(msgLicenseExpressionContainsInvalidCharacter,
+                                                  msg::value = static_cast<char>(cur())));
+                            next();
+                            break;
+                        }
+                        eat_idstring(result, expecting);
+                        break;
                 }
             }
 
-            if (!check_current_word())
+            if (expecting == Expecting::License)
             {
-                return nullopt;
+                add_error(msg::format(msgLicenseExpressionExpectLicenseFoundEof));
             }
-            else
+            if (expecting == Expecting::Exception)
+            {
+                add_error(msg::format(msgLicenseExpressionExpectExceptionFoundEof));
+            }
+
+            return result;
+        }
+    };
+
+    std::string parse_spdx_license_expression(StringView sv, Parse::ParseMessages& messages)
+    {
+        auto parser = SpdxLicenseExpressionParser(sv, "<license string>");
+        auto result = parser.parse();
+        messages = parser.extract_messages();
+        return result;
+    }
+
+    struct LicenseExpressionDeserializer : Json::IDeserializer<std::string>
+    {
+        virtual StringView type_name() const override { return "an SPDX license expression"; }
+
+        virtual Optional<std::string> visit_null(Json::Reader&) override { return {std::string()}; }
+
+        // if `sv` is a valid SPDX license expression, returns sv,
+        // but with whitespace normalized
+        virtual Optional<std::string> visit_string(Json::Reader& r, StringView sv) override
+        {
+            auto parser = SpdxLicenseExpressionParser(sv, "<manifest>");
+            auto res = parser.parse();
+
+            for (const auto& warning : parser.messages().warnings)
             {
-                return sv.to_string();
+                msg::println(Color::warning, warning.format("<manifest>", Parse::MessageKind::Warning));
+            }
+            if (auto err = parser.get_error())
+            {
+                r.add_generic_error(type_name(), err->format());
+                return nullopt;
```

Comment:
```suggestion
                return std::string{};
```
If the error has been handled, the Deserializer should not return `nullopt`. `nullopt` is reserved for (essentially) "method completely unimplemented, the parent must emit an error for me".

For example, `LicenseExpressionDeserializer::visit_object()` always returns `nullopt` and does not call `r.add_x_error()`, so the caller is responsible for adding an error.